habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py

```diff
@@ -315,3 +320,425 @@ def get_all_object_ids(sim: habitat_sim.Simulator) -> Dict[int, str]:
+
+    :return: A list of RaycastResults, one from each object keypoint.
+    """
+
+    if direction is None:
+        # default to downward raycast
+        direction = mn.Vector3(0, -1, 0)
+
+    global_keypoints = get_object_global_keypoints(objectA)
+    return [
+        sim.cast_ray(habitat_sim.geo.Ray(keypoint, direction))
+        for keypoint in global_keypoints
+    ]
+
+
+def get_object_set_from_id_set(
+    sim: habitat_sim.Simulator,
+    id_set: List[int],
+    ao_link_map: Optional[Dict[int, int]] = None,
+) -> List[
+    Union[
+        habitat_sim.physics.ManagedRigidObject,
+        habitat_sim.physics.ManagedArticulatedObject,
+    ]
+]:
+    """
+    Get the ManagedObjects from a set of object_ids.
+
+    :param sim: The Simulator instance.
+    :param id_set: The set of object ids for which ManagedObjects are desired.
+    :param ao_link_map: A pre-computed map from link object ids to their parent ArticulatedObject's object id.
+
+    :return: a list of tuples, first element is a ManagedObject, second is an optional link index.
+    """
+
+    if ao_link_map is None:
+        # Note: better to pre-compute this and pass it around
+        ao_link_map = get_ao_link_id_map(sim)
+
+    rom = sim.get_rigid_object_manager()
+    aom = sim.get_articulated_object_manager()
+    rigids = [
+        (rom.get_object_by_id(ro_id), None)
+        for ro_id in id_set
+        if rom.get_library_has_id(ro_id)
+    ]
+    aos = [
+        (
+            aom.get_object_by_id(ao_link_map[ao_id]),
+            aom.get_object_by_id(ao_link_map[ao_id]).link_object_ids[ao_id],
+        )
+        for ao_id in id_set
+        if ao_id in ao_link_map
+    ]
+
+    return rigids + aos
+
+
+def above(
+    sim: habitat_sim.Simulator,
+    objectA: habitat_sim.physics.ManagedRigidObject,
+    ao_link_map: Optional[Dict[int, int]] = None,
+) -> List[
+    Union[
+        habitat_sim.physics.ManagedRigidObject,
+        habitat_sim.physics.ManagedArticulatedObject,
+    ]
+]:
+    """
+    Get a list of all objects that a particular objectA is 'above'.
+    Concretely, 'above' is defined as: a downward raycast of any object keypoint hits the object below.
+
+    :param sim: The Simulator instance.
+    :param objectA: The ManagedRigidObject for which to query the 'above' set.
+    :param ao_link_map: A pre-computed map from link object ids to their parent ArticulatedObject's object id.
+
+    :return: a list of tuples, first element is a ManagedObject, second is an optional link index.
+    """
+
+    # get object ids of all objects below this one
+    above_object_ids = [
+        hit.object_id
+        for keypoint_raycast_result in object_keypoint_cast(sim, objectA)
+        for hit in keypoint_raycast_result.hits
+    ]
+    above_object_ids = list(set(above_object_ids))
+    return get_object_set_from_id_set(sim, above_object_ids, ao_link_map)
+
+
+def within(
```

Comment:
same question/suggestion as I had for the above(). Would be great to query it using objects instance names.